dist-tracing/src/trace.rs

```diff
@@ -0,0 +1,255 @@
+    pub instance_id: u64,
+}
+
+// TODO: round trip property test for this
+impl SpanId {
+    /// attempts to parse a SpanId from a string
+    pub fn from_string(s: &str) -> Option<SpanId> {
+        let mut iter = s.split('-');
+        let s1 = iter.next()?;
+        let u1 = u64::from_str_radix(s1, 10).ok()?;
+        let s2 = iter.next()?;
+        let u2 = u64::from_str_radix(s2, 10).ok()?;
+
+        Some(SpanId {
+            tracing_id: tracing::Id::from_u64(u1),
+            instance_id: u2,
+        })
+    }
+
+    /// convert a SpanId to a string that can be included in headers, RPC framework metadata fields, etc
+    pub fn to_string(&self) -> String {
+        format!("{}", self)
+    }
+}
+
+impl std::fmt::Display for SpanId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}-{}", self.tracing_id.into_u64(), self.instance_id)
+    }
+}
+
+/// NOTE: why not just have this be a u128? It'd be so much better...
+/// A Honeycomb Trace ID. Uniquely identifies a single distributed (potentially multi-process) trace.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct TraceId(pub(crate) u128);
+
+impl TraceId {
+    /// attempts to parse a TraceId from a string
+    pub fn from_string(s: &str) -> Option<Self> {
+        let u = u128::from_str_radix(s, 10).ok()?;
+        Some(TraceId(u))
+    }
+
+    /// convert a TraceId to a string that can be included in headers, RPC framework metadata fields, etc
+    pub fn to_string(&self) -> String {
+        format!("{}", self.0)
+    }
+
+    /// Generate a random trace ID by using a thread-level RNG to generate a u128
+    pub fn generate() -> Self {
+        use rand::Rng;
+        let u: u128 = rand::thread_rng().gen();
+
+        TraceId(u)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Span<'a, V> {
+    pub id: SpanId,
+    pub trace_id: TraceId,
+    pub parent_id: Option<SpanId>,
+    pub initialized_at: DateTime<Utc>,
+    pub elapsed_ms: i64,
+    pub level: tracing::Level,
+    pub name: &'a str,
+    pub target: &'a str,
+    pub service_name: &'a str,
+    pub values: V, // visitor used to record values
+}
+
+impl<'a, V> Span<'a, V> {
+    #[cfg(test)]
+    pub(crate) fn into_static(self) -> Span<'static, V> {
+        Span {
+            name: test::lift_to_static(self.name),
+            target: test::lift_to_static(self.target),
+            service_name: test::lift_to_static(self.service_name),
+            id: self.id,
+            trace_id: self.trace_id,
+            parent_id: self.parent_id,
+            initialized_at: self.initialized_at,
+            elapsed_ms: self.elapsed_ms,
+            level: self.level,
+            values: self.values,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Event<'a, V> {
+    pub trace_id: TraceId,
+    pub parent_id: Option<SpanId>,
+    pub initialized_at: DateTime<Utc>,
+    pub level: tracing::Level,
+    pub name: &'a str,
+    pub target: &'a str,
```

Comment:
these are all parts of the event's metadata, and `Event::metadata` returns a static reference. you could just replace these fields with a `&'static Metadata<'static>`, saving two wide ptrs worth of space